### PR TITLE
Nio

### PR DESCRIPTION
**Pull request by pedantix**
@tanner0101 I don't think you should merge this onto beta directly, I just didn't have a branch to merge it onto. Well, this is my first pass at the Redis implementation.

### CODE REVIEW

File: Sources/Redis/Client/RedisClient+Channel.swift [hunk 1]
```diff
@@ -1,24 +1,41 @@
...
 
 extension RedisClient {
-    /// Publishes data to a Redis channel.
-    public func publish(_ message: RedisData, to channel: String) -> Future<RedisData> {
-        return command("PUBLISH", [.bulkString(channel), message])
+    /// Subscribes to the channels and call the subscription handler on message from server
+    /// Should be closed when subscription not needed.
+    /// Makes the redis client subscription only.
+    public func subscribe(_ channels: Set<String>,
+                          subscriptionHandler: @escaping (RedisChannelData) -> Void) throws -> Future<Void> {
```
Comment (tanner0101) [inline Sources/Redis/Client/RedisClient+Channel.swift]: all Xcode-aided formatting must be abolished from this PR

File: Sources/Redis/Client/RedisClient+KeyedCache.swift [hunk 2]
```diff
@@ -1,20 +1,31 @@
+//
+//  RedisClient+KeyedCache.swift
+//  Async
+//
+//  Created by Shaun Hubbard on 3/9/18.
```
Comment (tanner0101) [inline Sources/Redis/Client/RedisClient+KeyedCache.swift]: no file headers. (your name will be in the commits and as a contributor on gh :) )

File: Sources/Redis/Client/RedisClient+connect.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import NIO
+
+extension RedisClient {
+    /// Connects to a Redis server using a TCP socket.
+    public static func connect(
+        hostname: String = "127.0.0.1",
+        port: Int = 6379,
+        on worker: Worker,
+        onError: @escaping (Error) -> Void
+        ) throws -> Future<RedisClient> {
```
Comment (tanner0101) [inline Sources/Redis/Client/RedisClient+connect.swift]: `)` line must be indented to the left
(xcode is so annoying about this)

File: Sources/Redis/Client/RedisClient.swift [hunk 4]
```diff
@@ -39,25 +47,30 @@ public final class RedisClient {
...
 
 /// MARK: Config
 
 /// Config options for a `RedisClient.
-struct RedisClientConfig: Codable {
+public struct RedisClientConfig: Codable {
     /// Default `RedisClientConfig`
     public static func `default`() -> RedisClientConfig {
-        return .init(hostname: "localhost", port: 6379)
+        return .init(hostname: "127.0.0.1", port: 6379)
```
Comment (tanner0101) [inline Sources/Redis/Client/RedisClient.swift]: we should keep this `localhost`, as that's how it appears in all other packages

File: Tests/RedisTests/RedisDataDecoderTests.swift [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        try bulkStringTestCase(protocolString: "$1\r\na\r\n", expectedString: "a")
+        try assertWillParseNil(string: "$3\r\naaa\r")
+        try bulkStringTestCase(protocolString: "$3\r\nfoo\r\n",
+                               expectedString: "foo")
+        try bulkStringTestCase(protocolString: "$3\r\naaa\r\n", expectedString: "aaa")
+        try bulkStringTestCase(protocolString: "$1\r\na\r\n$2\r\naa\r\n",
+                               expectedString: "a",
+                               otherString: "aa")
+
+        // decode special charachters guard against multibyte decoding
```
Comment (tanner0101) [inline Tests/RedisTests/RedisDataDecoderTests.swift]: nice tests :)

**Comment by tanner0101** [comment]
We should have this on an `nio` branch like all the other repos. You should have permissions to create that

File: Sources/Redis/Client/RedisClient.swift (hunk 4 above)
Comment (pedantix) [reply Sources/Redis/Client/RedisClient.swift]: I was getting connection issues with localhost. I will try it again

File: Tests/RedisTests/RedisDataDecoderTests.swift (hunk 5 above)
Comment (pedantix) [reply Tests/RedisTests/RedisDataDecoderTests.swift]: I tried... the tests for implementing `AMQP` are going to be rediculous

File: Sources/Redis/Client/RedisClient+Channel.swift (hunk 1 above)
Comment (pedantix) [reply Sources/Redis/Client/RedisClient+Channel.swift]: I am afraid this was an artifact of SwiftLint formatting.  I will go ahead and put all the function bodies in a  single line. to remove the formatting.

File: Sources/Redis/Client/RedisClient+KeyedCache.swift (hunk 2 above)
Comment (pedantix) [reply Sources/Redis/Client/RedisClient+KeyedCache.swift]: D'oh... Xcode... I really should take some time and write a VaporPlugin for Emacs....

Updated Priorities Lists 
1. VGSM - API -  Deployment to Vapor Cloud???
2. Vapor/Services - Draft Life Cycle RFC
3. Vapor/AMQP
4. VGSM Client(s) iOS/Web/Droid
5. Emacs Plugin For Vapor

File: Sources/Redis/Client/RedisClient+Channel.swift [hunk 6]
```diff
@@ -1,24 +1,43 @@
...
+    /// Maps RedisData.array to RedisChannelData, throws if map fails
+    private func convert(channelMessage: RedisData) throws -> RedisChannelData? {
+        guard let array = channelMessage.array,
+            array.first?.string == "message" // must contain ["message", <channel>, <redisData>]
+             else {
+            return nil
+        }
+        guard let channel = array[1].string else {
+            throw RedisError(identifier: "channel data",
+                             reason: "channel data did not contain identifier",
```
Comment (tanner0101) [inline Sources/Redis/Client/RedisClient+Channel.swift]: d'oh, still some persistent Xcode formatting

**Comment by tanner0101** [comment]
thanks @pedantix!

**Pull request merged**